srtcore/group.cpp

```diff
@@ -1918,6 +1918,218 @@ struct FLookupSocketWithEvent_LOCKED
+        // enough for SRT to maintain the connection, but it will still drop
+        // packets that didn't arrive in time), in both cases it may
+        // potentially block the reading for an indefinite time, while
+        // simultaneously a kangaroo might be a link that got some packets
+        // dropped, but then it's still capable to deliver packets on time.
+
+        // Note that gi->id might be a socket that was previously being polled
+        // on write, when it's attempting to connect, but now it's connected.
+        // This will update the socket with the new event set.
+
+        aliveMembers.push_back(gi->ps);
+        HCLOG(ds << "@" << gi->id << "[READ] ");
+    }
+
+    HLOGC(grlog.Debug, log << "group/recv: " << ds.str() << " --> EPOLL/SWAIT");
+#undef HCLOG
+
+    return aliveMembers;
+}
+
+vector<CUDTSocket*> CUDTGroup::recv_CollectReadReady(set<CUDTSocket*>& broken)
+{
+    // TODO: Function. Get a list of readable (non-broken) sockets.
+    {
+        // You can't lock the whole group for that
+        // action because this will result in a deadlock.
+        // Prepare first the list of sockets to be added as connect-pending
+        // and as read-ready, then unlock the group, and then add them to epoll.
+        vector<CUDTSocket*> aliveMembers = recv_CollectNonBroken(broken);
+        const bool still_alive = !aliveMembers.empty();
+
+        /* Done at the connecting stage so that it won't be missed.
+
+           int connect_modes = SRT_EPOLL_OUT | SRT_EPOLL_ERR;
+           for (vector<SRTSOCKET>::iterator i = connect_pending.begin(); i != connect_pending.end(); ++i)
+           {
+           XXX This is wrong code; this should use the internal function and pass CUDTSocket*
+           epoll_add_usock_INTERNAL(m_RcvEID, i->second, &connect_modes);
+           }
+
+           AND this below additionally for sockets that were so far pending connection,
+           will be now "upgraded" to readable sockets. The epoll adding function for a
+           socket that already is in the eid container will only change the poll flags,
+           but will not re-add it, that is, event reports that are both in old and new
+           flags will survive the operation.
+
+         */
+
+        for (vector<CUDTSocket*>::iterator i = aliveMembers.begin(); i != aliveMembers.end(); ++i)
+        {
+            // NOT using the official srt_epoll_add_usock because this will do socket dispatching,
+            // which requires lock on m_GlobControlLock, while this lock cannot be applied without
+            // first unlocking m_GroupLock.
+            const int read_modes = SRT_EPOLL_IN | SRT_EPOLL_ERR;
+            CUDT::s_UDTUnited.epoll_add_usock_INTERNAL(m_RcvEID, *i, &read_modes);
+        }
+
+        if (!still_alive)
+        {
+            LOGC(grlog.Error, log << "group/recv: all links broken");
+            throw CUDTException(MJ_CONNECTION, MN_NOCONN, 0);
+        }
```

Comment:
Here we can first check `if (!still_alive)` and throw an error if none alive.
Then loop through alive sockets and add to epoll.